modules/core/shared/src/main/scala/gem/math/Wavelength.scala

```diff
@@ -9,33 +9,97 @@ import gem.syntax.prism._
+ * @param toPicometers This wavelength in integral picometers (10^-12 of a meter).
  */
-sealed abstract case class Wavelength private (toAngstroms: Int) {
+sealed abstract case class Wavelength private (toPicometers: Int) {
   // Sanity check … should be correct via the companion constructor.
-  assert(toAngstroms >= 0, s"Invariant violated. $toAngstroms is negative.")
+  assert(toPicometers >= 0, s"Invariant violated. $toPicometers is negative.")
 }
 
 object Wavelength {
 
-  final lazy val Min: Wavelength = fromAngstroms.unsafeGet(0)
-  final lazy val Max: Wavelength = fromAngstroms.unsafeGet(Int.MaxValue)
+  final lazy val Min: Wavelength = fromPicometers.unsafeGet(0)
+  final lazy val Max: Wavelength = fromPicometers.unsafeGet(Int.MaxValue)
 
   /** @group Typeclass Instances */
   implicit val WavelengthShow: Show[Wavelength] =
     Show.fromToString
 
   /** @group Typeclass Instances */
   implicit val WavelengthOrd: Order[Wavelength] =
-    Order.by(_.toAngstroms)
+    Order.by(_.toPicometers)
 
   /**
-   * Prism from Int into Wavelength and back.
+   * Prism from Int in pm into Wavelength and back.
    * @group Optics
    */
-  def fromAngstroms: Prism[Int, Wavelength] =
-    Prism((n: Int) => Some(n).filter(_ >= 0).map(new Wavelength(_) {}))(_.toAngstroms)
+  def fromPicometers: Prism[Int, Wavelength] =
+    Prism((n: Int) => Some(n).filter(_ >= 0).map(new Wavelength(_) {}))(_.toPicometers)
+
+
+  // A Format such as this one which converts to Å:
+  //  fromPicometers.asFormat.imapA(_ / 100, _ * 100)
+  // is not a valid Format since λ => Int (Å) => λ loses precision
+
+  /**
+   * Poor man's uncomposable "optics" for converting between wavelength units.
+   * @param name
+   * @param exp power of 10 difference relative to pm (Å is 2, nm is 3, μm 6)
+   */
+  sealed abstract case class UnitConverter(name: String, exp: Int) {
+    private val p: Int = BigInt(10).pow(exp).toInt
+
+    /**
+     * Max value in these units that can be stored in a Wavelength.
+     */
+    val maxValue: Int =
+      Int.MaxValue / p
+
+    /**
+     * Creates a Wavelength from an Int in the corresponding units, provided it
+     * is in the range 0 to `maxValue`.
+     * @param n value in corresponding unit
+     * @return Some(Wavelength) provided n is in range [0, `maxValue`], None
+     *         otherwise
+     */
+    def getOption(n: Int): Option[Wavelength] =
+      if ((n < 0) || (n > maxValue)) None else fromPicometers.getOption(n * p)
+
+    def unsafeGet(n: Int): Wavelength =
+      getOption(n).getOrElse(sys.error(s"$n ($name) is not in range [0, $maxValue]"))
+
+    /**
+     * Converts Wavelength to an Int in the corresponding units, losing precision.
+     */
+    def reverseGet(w: Wavelength): Int =
+      w.toPicometers / p
+
+  }
+
+  /**
+   * Creates a Wavelength from an Int in Å.
+   * @param Å in the range (0 to 21474836)
+   * @return a Wavelength provided Å is in range
+   */
+  def fromAngstroms: UnitConverter =
```

Comment:
Should these be `unsafeFromAngstroms` given they use `UnitConverter`?

Reply:
It's always safe to produce the value as `Int` Å and there is a safe option for reading an `Int` value in Å so I think `fromAngstroms` is appropriate.  What is unsafe is assuming that an integer value can be represented in the `Wavelength` type.  If it is less than 0 or more than 21474836 we cannot represent it with an `Int` pm.  For that there is `fromAngstroms.unsafeGet` like for actual `Optics`.